LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/DefaultLoRaDataRequestHandler.cs

```diff
@@ -0,0 +1,505 @@
+                    macbytes = macCommandHolder.MacCommand[0].ToBytes();
+                }
+
+                if (cloudToDeviceMessage.Properties.TryGetValueCaseInsensitive("confirmed", out var confirmedValue) && confirmedValue.Equals("true", StringComparison.OrdinalIgnoreCase))
+                {
+                    requiresDeviceAcknowlegement = true;
+                    loRaDevice.LastConfirmedC2DMessageID = cloudToDeviceMessage.MessageId ?? Constants.C2D_MSG_ID_PLACEHOLDER;
+                }
+
+                if (cloudToDeviceMessage.Properties.TryGetValueCaseInsensitive("fport", out var fPortValue))
+                {
+                    fport = byte.Parse(fPortValue);
+                }
+
+                Logger.Log(loRaDevice.DevEUI, $"Sending a downstream message with ID {ConversionHelper.ByteArrayToString(rndToken)}", LogLevel.Debug);
+
+                frmPayload = cloudToDeviceMessage?.GetBytes();
+
+                Logger.Log(loRaDevice.DevEUI, $"C2D message: {Encoding.UTF8.GetString(frmPayload)}, id: {cloudToDeviceMessage.MessageId ?? "undefined"}, fport: {fport}, confirmed: {requiresDeviceAcknowlegement}, cidType: {cidTypeValue}", LogLevel.Information);
+
+                // cut to the max payload of lora for any EU datarate
+                if (frmPayload.Length > 51)
+                    Array.Resize(ref frmPayload, 51);
+
+                Array.Reverse(frmPayload);
+            }
+
+            if (fpending)
+            {
+                fctrl |= (int)FctrlEnum.FpendingOrClassB;
+            }
+
+            // if (macbytes != null && linkCheckCmdResponse != null)
+            //     macbytes = macbytes.Concat(linkCheckCmdResponse).ToArray();
+            var srcDevAddr = upstreamPayload.DevAddr.Span;
+            var reversedDevAddr = new byte[srcDevAddr.Length];
+            for (int i = reversedDevAddr.Length - 1; i >= 0; --i)
+            {
+                reversedDevAddr[i] = srcDevAddr[srcDevAddr.Length - (1 + i)];
+            }
+
+            var msgType = requiresDeviceAcknowlegement ? LoRaMessageType.ConfirmedDataDown : LoRaMessageType.UnconfirmedDataDown;
+            var ackLoRaMessage = new LoRaPayloadData(
+                msgType,
+                reversedDevAddr,
+                new byte[] { fctrl },
+                BitConverter.GetBytes(fcntDown),
+                macbytes,
+                fport.HasValue ? new byte[] { fport.Value } : null,
+                frmPayload,
+                1);
+
+            // var firstWindowTime = timeWatcher.GetRemainingTimeToReceiveFirstWindow(loRaDevice);
+            // if (firstWindowTime > TimeSpan.Zero)
+            //     System.Threading.Thread.Sleep(firstWindowTime);
+            var receiveWindow = timeWatcher.ResolveReceiveWindowToUse(loRaDevice);
+            if (receiveWindow == 0)
+                return null;
+
+            string datr;
+            double freq;
+            long tmst;
+            if (receiveWindow == 2)
```

Comment:
would be cool to have that magical number 2 defined as a constant that can give some context to its meaning.

Reply:
I will add RECEIVE_WINDOW_1 and RECEIVE_WINDOW_2 as constants.